rule index_bam:
	input: 
		"mapping/{reference}/{aligner}/{unit}/pass_2_output/Aligned.sortedByCoord.out.bam"
	output: 
		"mapping/{reference}/{aligner}/{unit}/pass_2_output/Aligned.sortedByCoord.out.bam.bai"
	shell: 
		"/usr/local/bin/samtools index {input}"

rule star_mapping_pass2_run:
	input:
		dummy = "mapping/{reference}/{aligner}/{unit}/pass_2_index/SAindex",
		fq = lambda wildcards: config["units"][wildcards.unit]
	output:
		"mapping/{reference}/{aligner}/{unit}/pass_2_output/Aligned.sortedByCoord.out.bam"
	threads: 32
	params: 
		pass2_index_path = "mapping/{reference}/{aligner}/{unit}/pass_2_index/",
		pass2_run_path = "mapping/{reference}/{aligner}/{unit}/pass_2_output/"
	shell:"""
	mkdir -p {params.pass2_run_path}
	/home/alex.casella/software/STAR --genomeDir {params.pass2_index_path} --runThreadN {threads} --readFilesIn {input.fq} --readFilesCommand zcat --chimSegmentMin {config[chimSegmentMin]} --outFileNamePrefix {params.pass2_run_path} --outSAMtype BAM SortedByCoordinate
	"""

rule star_mapping_pass2_index:
	input:
		reference = lambda wildcards: config["references"][wildcards.reference],
		sjtab = "mapping/{reference}/{aligner}/{unit}/pass_1_output/SJ.out.tab"
	output:
		"mapping/{reference}/{aligner}/{unit}/pass_2_index/SAindex"
	threads: 32
	params: 
		path = "mapping/{reference}/{aligner}/{unit}/pass_2_index/"
	shell:"""
	mkdir -p {params.path}
	/home/alex.casella/software/STAR --runMode genomeGenerate --genomeDir {params.path} --genomeFastaFiles {input.reference} --sjdbFileChrStartEnd {input.sjtab} --sjdbOverhang {config[sjdbOverhang]} --runThreadN {threads}
	"""
    		
rule star_mapping:
	input:
		dummy = "mapping/{reference}/{aligner}/{unit}/pass_1_index/SAindex",
		fq = lambda wildcards: config["units"][wildcards.unit]
	output:
		bam = "mapping/{reference}/{aligner}/{unit}/pass_1_output/Aligned.sortedByCoord.out.bam",
		sjtab = "mapping/{reference}/{aligner}/{unit}/pass_1_output/SJ.out.tab"
	params:
		out_path = "mapping/{reference}/{aligner}/{unit}/pass_1_output/",
		index_dir = "mapping/{reference}/{aligner}/{unit}/pass_1_index/"
	log:
		"mapping/{reference}/{aligner}/{unit}/log/{unit}_star_mapping.log"
	threads: 32
	shell:"""
	mkdir -p {params.out_path}
	/home/alex.casella/software/STAR --genomeDir {params.index_dir} --runThreadN {threads} --readFilesIn {input.fq} --readFilesCommand zcat --chimSegmentMin {config[chimSegmentMin]} --outFileNamePrefix {params.out_path} --outSAMtype BAM SortedByCoordinate
	"""
	
rule star_mapping_pass1_index:
	input:
		reference = lambda wildcards: config["references"][wildcards.reference],
		gtf = lambda wildcards: config["annotation_gtf"]
	output:
		"mapping/{reference}/{aligner}/{unit}/pass_1_index/SAindex"
	threads: 32
	params: 
		path = "mapping/{reference}/{aligner}/{unit}/pass_1_index/"
	shell:"""
	mkdir -p {params.path}
	/home/alex.casella/software/STAR --runMode genomeGenerate --genomeDir {params.path} --genomeFastaFiles {input.reference} --runThreadN {threads} --sjdbGTFfile {input.gtf} --sjdbOverhang {config[sjdbOverhang]}
	"""